### ISSUE TITLE
Concurrently replicating multiple db's in the browser with go-ipfs and ipfs-http-client

### ISSUE THREAD

**Issue by Loosetooth** [orbitdb/orbitdb#875]
I'm currently having a problem replicating multiple db's at the same time in the browser.
### The problem:
In certain circumstances, replicating multiple db's in the browser fails. The code infinitely hangs on `db.load()`. This depends on the number of databases being replicated, and the number of nodes hosting these databases. In the last few days I did a few tests trying to consistently replicate the problem. I'll try to briefly explain the different situations that work or fail here, as well as the used setup.

### Setup
I used a go-ipfs node together with the ipfs-http-client. The go-ipfs node has [a few changed settings](https://stackoverflow.com/questions/42708251/how-to-do-cross-origin-requests-on-ipfs) to circumvent a CORS error when connecting to the node from the browser, and is run with the `--enable-pubsub-experiment` flag. I tried js-ipfs before, but didn't manage to get peer discovery working at all. Using go-ipfs, peers are easily discovered on the local network.

I then continued to run multiple nodes/orbitdb-instances on my local network, each creating/hosting at least one unique database. Every node was run on a separate computer. Hosting the databases was done in both nodejs and the browser, this seemed to always work, so I suppose this has no effect on the problem. The next step was to try to sync all the databases on another, separate, computer. Replicating the databases was tried in nodejs and in the browser. Playing with the number of databases and nodes results in different behavior:

#### Works:
In nodejs:
- Replicating in nodejs always works. Regardless of whether the databases are hosted in the browser or in nodejs.

In the browser:
- Replicating 2 db's or less. (From a single or multiple nodes.)
- Replicating up to 4 db's from a single node.
- Replicating 1 db from node A and 2 db's from node B.

#### Fails (in the browser):
- Replicating more than 4 db's from a single node.
- Replicating 1 db from node A and 3 or more db's from node B.
- Replicating 3 db's (one on each node) from nodes A, B and C.
- Replicating 2 db's from node A and 2 db's from node B (this also fails with more db's or nodes) 

#### Failure details:
When the db has no data that is locally synced, the `db.load()` line completes, but the database stays empty. (No synchronization.) If any data has already been locally synced (before), the code will hang infinitely on `db.load()`.

I made an [example react project](https://github.com/Loosetooth/orbitdbbugreact) showcasing the problem, since it's relatively complex to replicate. You can run `node src/bug.js` to create a certain amount of db's or try to sync a list of db addresses in nodejs. Or you can `yarn start` to replicate a list of db's in the browser. (Paste the list in `index.js`.)

**Comment by tabcat** [comment]
so it seems like you bring up 2 issues; an issue with replicating a certain amount of dbs from other nodes into the browser and an issue with the `db.load` function hanging. these sound familiar and hopefully are not difficult to fix and test.

its super awesome you made a repo to reproduce the bugs so thank you for that. I will definitely run this when I get the chance, follow up, and then see what can be done to remedy these issues. :+1: 

in the meantime feel free to also bring this up in the community chat if you havent already. [gitter](https://gitter.im/orbitdb/Lobby) [matrx](https://riot.im/app/#/room/#orbit-db:matrix.org)

**Comment by Loosetooth** [comment]
Yes, you're correct about the two separate issues. I rather thought of it as "synchronization problems", but splitting it up like this probably helps to recognize the problems faster. Thanks!

**Comment by tabcat** [comment]
> When the db has no data that is locally synced, the db.load() line completes, but the database stays empty. (No synchronization.) If any data has already been locally synced (before), the code will hang infinitely on db.load().

I'm wondering if this is a result of referenced data being garbage collected. if you are using go-ipfs, it does garbage collect and orbitdb does not pin entries by default. if entries are garbage collected and dont exist somewhere else they are practically lost.
to clear up the hanging issue for this you can [give `.load` an option called `fetchEntryTimeout`](https://github.com/orbitdb/orbit-db-store/blob/11d50dd66f08487356c4b97cccdbe16bd047cee4/src/Store.js#L228-L234) which should resolve that.
however you are still losing entries locally which is not ideal in most situations. there is an [option to pin entries in the store superclass](https://github.com/orbitdb/orbit-db-store/blob/11d50dd66f08487356c4b97cccdbe16bd047cee4/src/Store.js#L417-L425) and any of the orbitdb write methods should pass their options through to that. pinning may affect performance.

**Comment by Loosetooth** [comment]
I tried your ideas in the "tabcatFix" class in the testrepository. (Both timeout and pinning.) Didn't work, fetching like 8 db's from a local node only syncs the first few, then hangs forever on the `.load()` of one of the later db's. The *fetchEntryTimeout* on its own also doesn't seem to have any effect.

Perhaps my description of the problem details was wrong... The first few db's seem to load anyway. Sorry for that mistake in the description above. The later databases still hang on load() anyway though!

**Comment by phillmac** [comment]
@tabcat theres a bug in fetchEntryTimeout, half of the fetch ops ignore it. 
@Loosetooth does it always hang on the same db? What if you load only that db on its own?
Its possible to hang the ipfs node too by overworking it. (too much disk io to handle any requests in a timely maner) so id also check if ipfs is reponding outside of the app, e.g. `ipfs dag fetch` something

**Comment by tabcat** [comment]
if you close one store is another store able take its place and replicate/load after?

> Replicating in nodejs always works. Regardless of whether the databases are hosted in the browser or in nodejs.

so this is specific to using go-ipfs via the http api?

**Comment by Loosetooth** [comment]
> does it always hang on the same db? What if you load only that db on its own?

@phillmac Yes, when loading the same series of db's, it usually hangs on the same one. In the case where I tried with 8 db's it always hung on the `.load()` of the 6th db. When I load that specific db on its own it loads up just fine.

> Its possible to hang the ipfs node too by overworking it...

I will check if this is the case later! Thanks for the idea.

> if you close one store is another store able take its place and replicate/load after?

@tabcat Yes, I have tried sequentially loading the db's and closing them again, one after another. This seems to work fine. So the problem is specific to opening multiple db's at the same time.

>> Replicating in nodejs always works. Regardless of whether the databases are hosted in the browser or in nodejs.

> so this is specific to using go-ipfs via the http api?

No, when running the example code in nodejs, the http-ipfs-client is still used. This works fine, so it eliminates the combination of go-ipfs/http-api being the core problem.

P.S. I'm doing some more tests using electron, and it seems to have the same problems as in the browser. (When running in the main.js/nodejs part of electron.) Maybe this can be a hint to where the problem lies, but let's not focus on that here yet.

**Comment by tabcat** [comment]
so just to be clear; the issue is with browser nodes using an ipfs http api that is talking to an instance of go-ipfs. is this correct?

**Comment by Loosetooth** [comment]
> so just to be clear; the issue is with browser nodes using an ipfs http api that is talking to an instance of go-ipfs. is this correct?

Yes, the http-ipfs-client can be run either in nodejs or the browser. The above problem only appears in the browser. The ipfs node that the client connects to is always a go-ipfs instance.

**Comment by Loosetooth** [comment]
@phillmac Just tested if the go-ipfs node was responsive with `ipfs cat ` while it was stuck on loading a db, and it responded just fine.

I also just noticed that this error is spammed like crazy in the browser if I wait for a long enough time while a db hangs on the `load` call: (This really takes a while to appear though!)

![image](https://user-images.githubusercontent.com/66420755/114273514-971cb400-9a1a-11eb-8f23-5456a27b243b.png)

**Comment by phillmac** [comment]
Just out of curiosity, can you post a .har of the network graph? Devtools
=> network => export?

Sounds a bit like this issue:

https://bugs.chromium.org/p/chromium/issues/detail?id=108055

```
This is almost certainly the net::ERR_INSUFFICIENT_RESOURCES line
ResourceDispatcherHostImpl cancelling requests when a renderer uses too
much browser memory.

Solution is probably to hold back (low-priority?) requests in the renderer
process when there are too many pending requests. Not strictly a network
issue. Think we should keep this one around, though I suspect a fix may be
fairly involved.
```

**Comment by phillmac** [comment]
So the next thing to test is instead of simply awaiting load() which will
return almost instantly when there is no replication history, resulting  in
all the dbs being queued for replication simultaneously, maybe use a
replicated event to limit the number concurent dbs? That is, wait for one
or two to finish before even trying to open the next.
Maybe a simple work around for the purpose of testing would be to introduce
an artificial 30 sec delay after the await load() and see if that improves
the symptoms at all.

**Comment by phillmac** [comment]
Also, I'm assuming that you're using chrome, so definitely try firefox to
see if this is a chrome specific bug.

**Comment by Loosetooth** [comment]
First of all, thanks again for the help! It's very much appreciated.

> Maybe a simple work around for the purpose of testing would be to introduce an artificial 30 sec delay after the await load() and see if that improves the symptoms at all

I made a [*delayedLoad* branch](https://github.com/Loosetooth/orbitdbbugreact/tree/delayedLoad) in the test repository doing what you proposed. I tried to load 8 databases from a node in the local network. I tried with different delay times ranging from 1s to 30s, which had no effect on the problem. The 6th db always hangs on the `.load()` call, and after the errors start spamming.

I tried a 3s delay between loading the db's (in the brave browser) and recorded a .har file:
[localhost.har.zip](https://github.com/orbitdb/orbit-db/files/6291489/localhost.har.zip)

The reason I took 3s delay for the recording instead of 30s was that the .har file gets corrupted once the errors start spamming, and it's a little harder to time with 30s delay. But it results in the same behavior so I don't think it should make a difference.

> Also, I'm assuming that you're using chrome, so definitely try firefox to see if this is a chrome specific bug.

I ran it in firefox too, gives the same behavior. Gets stuck on the 6th `db.load()`. However: in firefox the console is not spammed full of errors, instead the whole browser becomes unresponsive after a while. So this does not seem to be a chrome specific bug.
